File: HoTT-MdlViewer-Swing/src/main/java/gde/mdl/ui/SelectFromMemory.java
/**
 *  HoTT Transmitter Config
 *  Copyright (C) 2013  Oliver Treichel
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package gde.mdl.ui;

import gde.model.BaseModel;
import gde.model.enums.ModelType;
import gde.model.serial.ModelInfo;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import javax.swing.AbstractListModel;
import javax.swing.JComponent;
import javax.swing.JList;
import javax.swing.ListSelectionModel;
import javax.swing.SwingWorker;

import de.treichels.hott.HoTTDecoder;

/**
 * @author oli@treichels.de
 */
public class SelectFromMemory extends SelectFromTransmitter {
  private class ArrayListModel extends AbstractListModel<String> {
    private static final long     serialVersionUID = 1L;
    private final List<ModelInfo> modelInfos       = new ArrayList<ModelInfo>();

    public void addModelInfo(final ModelInfo info) {
      if (info.getModelType() != ModelType.Unknown) {
        final int index = getSize();
        modelInfos.add(info);
        fireIntervalAdded(this, index, index);
      }
    }

    public void clear() {
      modelInfos.clear();
    }

    @Override
    public String getElementAt(final int index) {
      final ModelInfo info = modelInfos.get(index);

      return String.format("%02d: %c%s.mdl", info.getModelNumber(), info.getModelType() == ModelType.Helicopter ? 'h' : 'a', info.getModelName()); //$NON-NLS-1$
    }

    public int getModelNumerAt(final int index) {
      return modelInfos.get(index).getModelNumber();
    }

    @Override
    public int getSize() {
      return modelInfos.size();
    }
  }

  private final class ReloadWorker extends SwingWorker<Void, ModelInfo> {
    @Override
    protected Void doInBackground() throws Exception {
      layerUI.start();
      list.setEnabled(false);
      final ModelInfo infos[] = port.getAllModelInfos();

      for (final ModelInfo info : infos) {
        publish(info);
      }

      return null;
    }

    @Override
    protected void done() {
      layerUI.stop();
      list.setEnabled(true);
    }

    @Override
    protected void process(final List<ModelInfo> chunks) {
      for (final ModelInfo info : chunks) {
        model.addModelInfo(info);
      }
    }
  }

  private static final long    serialVersionUID = 1L;

  private final ArrayListModel model            = new ArrayListModel();
  private final JList<String>  list             = new JList<String>(model);
  private int                  selectedIndex    = -1;

  public SelectFromMemory() {
    list.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);
    list.setLayoutOrientation(JList.VERTICAL);
    list.setVisibleRowCount(-1);

    initUI();

    onReload();
  }

  @Override
  public BaseModel getModel() throws IOException {
    if (selectedIndex == -1) {
      return null;
    }

    return HoTTDecoder.decodeMemory(port, model.getModelNumerAt(selectedIndex));
  }

  @Override
  protected JComponent getSelectionComponent() {
    return list;
  }

  @Override
  public void onCancel() {
    selectedIndex = -1;
  }

  @Override
  public void onOpen() {
    selectedIndex = list.getSelectedIndex();
  }

  @Override
  public void onReload() {
    selectedIndex = -1;
    model.clear();

    if (port != null) {
      new ReloadWorker().execute();
    }
  }
}

File: MdlViewer/src/main/kotlin/de/treichels/hott/mdlviewer/javafx/SelectFromMemory.kt
/**
 * HoTT Transmitter Config Copyright (C) 2013 Oliver Treichel
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program. If not, see <http:></http:>//www.gnu.org/licenses/>.
 */
package de.treichels.hott.mdlviewer.javafx

import de.treichels.hott.decoder.HoTTTransmitter
import de.treichels.hott.model.enums.ModelType
import de.treichels.hott.serial.ModelInfo
import javafx.beans.binding.BooleanBinding
import javafx.concurrent.Task
import javafx.scene.control.ListCell
import javafx.scene.control.ListView
import javafx.scene.control.SelectionMode
import javafx.util.Callback
import tornadofx.*

/**
 * @author Oliver Treichel &lt;oli@treichels.de&gt;
 */
class SelectFromMemory : SelectFromTransmitter() {
    private var listView by singleAssign<ListView<ModelInfo>>()

    // dialog is ready when the user selects an item from the list
    override fun isReady(): BooleanBinding = listView.selectionModel.selectedItemProperty().isNotNull

    // load the selected item as Model
    override fun getResult(): Task<Model>? {
        val modelInfo = listView.selectionModel.selectedItem
        val serialPort = this.transmitter

        return if (modelInfo != null && serialPort != null) {
            runAsync { Model.loadModel(modelInfo, serialPort) }
        } else null
    }

    init {
        title = messages["select_from_memory_title"]

        // add to UI
        root.center {
            listView = listview {
                selectionModel.selectionMode = SelectionMode.SINGLE
                setOnMouseClicked { handleDoubleClick(it) }
                cellFactory = Callback { ModelInfoListCell() }
            }
        }
    }

    /**
     * Refresh a [ListView] with a list of all models from the transmitter memory using the [HoTTTransmitter] from parent.
     */
    override fun refreshUITask(): Task<*> {
        // add temporary placeholder
        listView.items = observableList(null)

        // fill list in background
        return listView.runAsyncWithOverlay {
            transmitter?.allModelInfos?.filter { it.modelType != ModelType.Unknown } ?: listOf()
        }.success { list ->
            listView.items = list.observable()
        }
    }

    private inner class ModelInfoListCell : ListCell<ModelInfo>() {
        override fun updateItem(item: ModelInfo?, empty: Boolean) {
            super.updateItem(item, empty)

            if (!empty) {
                text = if (item == null) {
                    messages["loading"]
                } else {
                    String.format("%02d: %c%s.mdl", item.modelNumber + 1, item.modelType.char, item.modelName)
                }
            }
        }
    }
}


